RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc

```diff
@@ -0,0 +1,162 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "DataFormats/PatCandidates/interface/MET.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+
+struct DeepMETCache {
+  std::atomic<tensorflow::GraphDef*> graph_def;
+};
+
+class DeepMETProducer : public edm::stream::EDProducer<edm::GlobalCache<DeepMETCache> > {
+public:
+  explicit DeepMETProducer(const edm::ParameterSet&, const DeepMETCache*);
+  void produce(edm::Event& event, const edm::EventSetup& setup) override;
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  // static methods for handling the global cache
+  static std::unique_ptr<DeepMETCache> initializeGlobalCache(const edm::ParameterSet&);
+  static void globalEndJob(DeepMETCache*);
+
+private:
+  const edm::EDGetTokenT<std::vector<pat::PackedCandidate> > pf_token_;
+  float norm_;
+  bool ignore_leptons_;
+  unsigned int max_n_pf_;
+
+  tensorflow::Session* session_;
+
+  std::unordered_map<int, int32_t> charge_embedding_;
+  std::unordered_map<int, int32_t> pdg_id_embedding_;
+};
+
+namespace {
+  float divide_and_rm_outlier(float val, float norm) {
+    float ret_val = val / norm;
+    if (ret_val > 1e6 || ret_val < -1e6)
+      return 0.;
+    return ret_val;
+  }
+}  // namespace
+
+DeepMETProducer::DeepMETProducer(const edm::ParameterSet& cfg, const DeepMETCache* cache)
+    : pf_token_(consumes<std::vector<pat::PackedCandidate> >(cfg.getParameter<edm::InputTag>("pf_src"))),
+      norm_(cfg.getParameter<double>("norm_factor")),
+      ignore_leptons_(cfg.getParameter<bool>("ignore_leptons")),
+      max_n_pf_(cfg.getParameter<unsigned int>("max_n_pf")) {
+  session_ = tensorflow::createSession(cache->graph_def);
```

Comment:
can be moved to the initializer list (before ```{```)

Reply:
Done.